cedar-policy-core/src/entities/json/err.rs

```diff
@@ -482,19 +482,87 @@ pub enum JsonDeserializationErrorContext {
 
 /// Type mismatch error (in terms of `SchemaType`)
 #[derive(Debug, Diagnostic, Error)]
-#[error("type mismatch: value was expected to have type {expected}: `{}`",
-    match .actual_val {
-        Either::Left(pval) => format!("{pval}"),
-        Either::Right(expr) => display_restricted_expr(expr.as_borrowed()),
-    }
+#[error("type mismatch: value was expected to have type {expected}, but it {mismatch_reason}: `{}`",
+    display_restricted_expr(.actual_val.as_borrowed()),
 )]
 pub struct TypeMismatchError {
     /// Type which was expected
-    pub expected: Box<SchemaType>,
-    /// Value which doesn't have the expected type; represented as either a
-    /// PartialValue or RestrictedExpr, whichever is more convenient for the
-    /// caller
-    pub actual_val: Either<PartialValue, Box<RestrictedExpr>>,
+    expected: Box<SchemaType>,
+    /// Reason for the type mismatch
+    mismatch_reason: TypeMismatchReason,
+    /// Value which doesn't have the expected type
+    actual_val: Box<RestrictedExpr>,
+}
+
+#[derive(Debug)]
+enum TypeMismatchReason {
+    /// We saw this dynamic type which is not compatible with the expected
+    /// schema type.
+    UnexpectedType(Type),
+    /// We saw a record type expression as expected, but it contains an
+    /// attribute we didn't expect.
+    UnexpectedAttr(SmolStr),
+    /// We saw a record type expression as expected, but it did not contain an
+    /// attribute we expected.
+    MissingRequiredAtr(SmolStr),
+    /// No further detail available.
+    None,
+}
+
+impl std::fmt::Display for TypeMismatchReason {
```

Comment:
This display impl could be a `derive(Error)` and then marking each case with `#[error("actually has type {ty}")]` etc